In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

# Importing df_prods

In [9]:
path = r'C:\Users\Apex PC\Achievement 4 project 29-04-2025\03 Scripts\Original Folder\products.csv'

In [11]:
path

'C:\\Users\\Apex PC\\Achievement 4 project 29-04-2025\\03 Scripts\\Original Folder\\products.csv'

In [13]:
df_prods = pd.read_csv(os.path.join(path), index_col = False)

# Importing df_ords

In [25]:
path = r'C:\Users\Apex PC\Achievement 4 project 29-04-2025\03 Scripts\Prepared Folder\orders_wrangled.csv'

In [27]:
path

'C:\\Users\\Apex PC\\Achievement 4 project 29-04-2025\\03 Scripts\\Prepared Folder\\orders_wrangled.csv'

In [29]:
df_ords = pd.read_csv(os.path.join(path), index_col = False)


In [55]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

# Creating a new data frame

In [31]:
df_test = pd.DataFrame()

# Creating a mixed type column

In [33]:
df_test['mix'] = ['a', 'b', 1, True]

In [35]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [37]:
df_test['mix'] = df_test['mix'].astype('str')

In [39]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [49]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [51]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [61]:
df_prods.shape  # to compare the number of rows 

(49693, 5)

In [63]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]  # creating a new dataframe

In [65]:
df_prods_clean.shape

(49677, 5)

# finding duplicates in df_prods

In [67]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [69]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [73]:
df_prods_clean.shape   # to check the current number of rows

(49677, 5)

# creating a new dataframe

In [75]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [77]:
df_prods_clean_no_dups.shape      # to check the current number of rows

(49672, 5)

In [79]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

the frequency check on the "order_days_of_week" column should result in a minimum of 0 and a maximum of 6, giving a total of exactly 7 values for the 7 days of the week, the 50th percentile, in this case, is 3. Also, no negative minimum value for "days_since_prior_order" and a maximum larger than 24 for "order_hour_of_day" as there are only 24 hours in a day.

# to check for mixed type data in df_ords

In [92]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [124]:
df_ords['days_since_prior_order'].dtype   # checking the data type


dtype('float64')

# no mixed type data was found

# checking for missing values

In [94]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

# Here the only column with missing values is the "days_since_prior_order" column, and it’s missing 206209 values.

# creating a subset of the dataframe containing only the missing values

In [98]:
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [100]:
df_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

# Addressing missing values in df_ords

In [104]:
df_ords.median()          # looking for the median

Unnamed: 0                1710541.0
order_id                  1710542.0
user_id                    102689.0
order_number                   11.0
orders_day_of_week              3.0
order_hour_of_day              13.0
days_since_prior_order          7.0
dtype: float64

I believe that that the missing values would make a lot of difference in the analysis so i won't be deleting them.

In [154]:
df_ords['days_since_prior_order'].fillna(0, inplace=True)



C:\Users\Apex PC\AppData\Local\Temp\ipykernel_16260\232264340.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_ords['days_since_prior_order'].fillna(0, inplace=True)


In [140]:
df_ords.isnull().sum()   # to check if missing values have been filled

Unnamed: 0                0
order_id                  0
user_id                   0
order_number              0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

# Finding duplicates in df_ords

In [142]:
df_dups = df_ords[df_ords.duplicated()]

In [144]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

# No duplicates found

In [146]:
path = r'C:\Users\Apex PC\Achievement 4 project 29-04-2025\03 Scripts\Prepared Folder'	# exporting data

In [148]:
path

'C:\\Users\\Apex PC\\Achievement 4 project 29-04-2025\\03 Scripts\\Prepared Folder'

In [150]:
df_ords.to_csv(os.path.join(path, 'orders_prepared.csv'))

In [152]:
df_prods.to_csv(os.path.join(path, 'products_prepared.csv'))